VTS/Core/Implementations/WebSocketNetCore/WebSocketNetCoreImpl.cs

```diff
@@ -0,0 +1,213 @@
+        private readonly ConcurrentQueue<Action> _responseQueue;
+        private bool _attemptReconnect;
+
+        private Action _onConnect = () => { };
+        private Action _onDisconnect = () => { };
+        private Action<Exception> _onError = (e) => { };
+
+        private string _url = "";
+        private readonly IVTSLogger _logger;
+
+        public WebSocketNetCoreImpl(IVTSLogger logger)
+        {
+            _logger = logger;
+            _intakeQueue = new ConcurrentQueue<string>();
+            _responseQueue = new ConcurrentQueue<Action>();
+        }
+
+        public string GetNextResponse()
+        {
+            _intakeQueue.TryDequeue(out var response);
+            return response!;
+        }
+
+        public bool IsConnecting()
+        {
+            return _socket is { State: WebSocketState.Connecting };
+        }
+
+        public bool IsConnectionOpen()
+        {
+            return _socket is { State: WebSocketState.Open };
+        }
+
+        public void Send(string message)
+        {
+            var buffer = Encoder.GetBytes(message);
+            _socket?.SendAsync(new ReadOnlyMemory<byte>(buffer), WebSocketMessageType.Text, true, default)
+                .ConfigureAwait(false)
+                .GetAwaiter()
+                .GetResult();
+        }
+
+        public void Start(string url, Action onConnect, Action onDisconnect, Action<Exception> onError)
+        {
+            _url = url;
+            _socket = new ClientWebSocket();
+            _logger.Log($"Attempting to connect to {_url}");
+            _socket.Options.KeepAliveInterval = TimeSpan.FromSeconds(10);
+            _onConnect = onConnect;
+            _onDisconnect = onDisconnect;
+            _onError = onError;
+
+            Task.Run<Task>(async () =>
+            {
+                try
+                {
+                    await _socket.ConnectAsync(new Uri(_url), CancellationToken.None);
+                }
+                catch (Exception e)
+                {
+                    _responseQueue.Enqueue(() =>
+                    {
+                        _logger.LogError($"[{_url}] - Socket error...");
+                        _logger.LogError($"'{e.Message}', {e}");
+                        _onError(e);
+                    });
+                    return;
+                }
+
+                _responseQueue.Enqueue(() =>
+                {
+                    _onConnect();
+                    _logger.Log($"[{_url}] - Socket open!");
+                    _attemptReconnect = true;
+                });
+
+                while (true)
+                {
+                    var result = await _socket.ReceiveAsync(CancellationToken.None);
+                    if (result.closeStatus == null)
+                    {
+                        _responseQueue.Enqueue(() =>
+                        {
+                            if (result is { buffer: { }, messageType: WebSocketMessageType.Text })
```

Comment:
This is not valid syntax in the version of .NET that Unity (circa LTS 2019) uses, which is .NET 2.0 Standard.

As an aside, I'm showing my hand here, but I've also never seen this syntax before. I took this to mean `(result.buffer != null && result.messageType ==  WebSocketMessageType.Text)` when editing this to get it to compile, is that correct?

Reply:
Yep, that's exactly how it works.